Easylist-based ad blocker

Helium314 opened the pull request:
The hosts-based ad blocker sometimes is not enough, so I'm working on a blocker that understands at least some of the AdBlockPlus rules used in easylist.

So far I have it basically working, but not in a state that is acceptable. (it is active when 'no ad blocking' is selected, and significantly slows down first start, there is no way to update, it might sometimes block the wrong things,...)

I have some ideas on how to improve it (see comments in the code), but what I will actually do depends on whether you are actually interested in having this implemented in the browser. (I would only do some small changes if it's just for myself)

jamal2362 commented:
Wow that's really great. The bugs can certainly be fixed somehow that are still in there.

Is it also possible to enter an easylist as url instead of only via the assets file?
Or is both possible ?

Helium314 commented:
Currently the list is only in assets, and the parsed list would not even update if you update the app (without resetting).

I first want to decide how to actually store the data, as this will require adjusting the code quite a bit. Probably extend the existing hosts database is a good idea, the current storage in shared preferences was just done to at least have persistent lists, without thinking how they would/should be used.

Slion commented:
It sounds like a feature I would like to see merged to the main branch yes. Though it's the first time I hear about Easylist so I would need to look into first. Once it is production quality and fits within the app architecture it can be merged.

Helium314 commented:
Easylist (and easyprivacy) are basically just commonly used lists for adblock / ublock origin.
Any other list using AdBlockPlus rules should work.

Unfortunately an important part of the rules is not (and will not be) implemented; if I understand correctly the _content filters_ are applied to the page content, so not usable with URL-based blocking.

Slion commented:
So is that a block list provider or a block list syntax or both?
Fulguris only has URL based ad-block for now. However I guess we could easily implement a JavaScript based ad-blocker, jsoup could help there I guess. Though I have hardly done any work with JS on Fulguris, maybe just something to extract the page meta and the theme color like a year ago. 
My understanding is that @jamal2362 already implemented such JS solutions on his fork, possibly taken from SmartCookieWeb.

For the implementation we may want to make sure there is no library available that could help us before doing the hardwork ourselves. I came across those mozilla android components the other day for instance. Not sure that's any relevant to our topic though.

I have yet to take a look at your draft.

jamal2362 commented:
@Slion 
Easylist blocks more effectively and you can also block .js from websites with it.

Helium314 commented:
> So is that a block list provider or a block list syntax or both?

Easylist is the provider, so maybe the title is misleading... Syntax is from adblock: https://adblockplus.org/en/filter-cheatsheet

> jamal2362/Styx#33

hmm... if the uBlock origin extension can somehow be used, this should be a much better solution than what I'm trying to do

jamal2362 commented:
The uBlock thing exists with Firefox, but whether that is so applicable to the WebView browser I don't know.

Slion commented:
> hmm... if the uBlock origin extension can somehow be used, this should be a much better solution than what I'm trying to do

Thus my earlier comment about not reinventing the wheel. If there is an open source solution out there we could easily integrate then we should seriously look into that rather than going our own way.

jamal2362 commented:
@Slion try a look here
https://github.com/CookieJarApps/SmartCookieWeb-Preview
or here
https://github.com/hazuki0x0/YuzuBrowser

Helium314 commented:
> The uBlock thing exists in Firefox, but whether that is so applicable to the WebView browser I don't know.

On desktop it's fine, but on android it slows my (rather old) phone down enormously

jamal2362 commented:
I think we should look at Yuzu Browser, he has implemented it the best so far. It also has filters and you can add multiple lists.

Slion commented:
> On desktop it's fine, but on android it slows my (rather old) phone down enormously
If it’s JavaScript solution that's to be expected.

Helium314 commented:
I had not heard of yuzu browser until now (event though I actually searched for existing implementations of adblock-style filters)
At first glance it looks like all the rules are implemented, and it "feels" fast.

Not sure if the content-based element selection is implemented... I guess not

jamal2362 commented:
Unfortunately, I don't know how exactly the browser handles the lists. I have not dealt with it so far. 
Only I know that the well blocks the pages and it's fast.
I have used the browser 1.5 years ago, so I knew that he works with filter lists.

sweb also works with filter lists I think but it is not so well implemented.
You can see it in his readme that he works with easylist.

Here the source:
https://github.com/alandau/sweb

Helium314 commented:
Hmm... no matter what implementation of those rules will be implemented, it might be good to have some advanced settings like in uBlock origin (left half):
![ubo](https://user-images.githubusercontent.com/43007630/120599400-bf87c400-c447-11eb-985f-0f98641f430b.png)

If you are not familiar with it:
there are 2 columns to change, left is global, right is for the current domain
\+ and - reflect the number of allowed/blocked requests (small detail, but _really_ useful)
grey color: no additional rule (just use normal blocking)
red color: block all
green color: allow all
dark color: setting explicitely set
light color: setting inherited
more info in advanced part of https://github.com/gorhill/uBlock/wiki/Quick-guide:-popup-user-interface#the-overview-panel

This overlaps a lot with the domain settings... and actually is not dependent on the type of adblocker used. So maybe should be discussed somewhere else than here?

jamal2362 commented:
Doesn't the whole thing pull extremely on the performance of the device?

I know this for the desktop, but is this also usable for mobile devices?

Helium314 commented:
> Doesn't the whole thing pull extremely on the performance of the device?

The whole uBlock yes, but I was thinking that something like this might be a good UI for selectively (un)blocking content, globally or per domain.
So for example you could use it to disable js globally, but set a noop rule (grey BG) on ebay.com, so only the normal adblock rules apply. 

The part that slows down everything is the blocking itself (my guess: mostly the part that needs to read through the page source code), and this would probably be taken from Yuzu, and thus not slowing down (so much).

jamal2362 commented:
@Helium314 
I think about it that it is simply enough to adopt the adblocker from yuzu browser. It is more effective than what is built into Fulguris or Styx. And this one also works with lists like easylist and co.

Your pull request already goes in this direction and that should be maintained.

jamal2362 commented:
Or you can extend your pull request and do something with it instead of copying it from another browser.

Helium314 commented:
I didn't find it that simple to use the blocker from Yuzu, but that's probably because I don't have much experience in this regard. And also because there are no comments in the Yuzu adblocker...

Anyway, I have it working now, but currently just for testing. Only one list, parsed every time after browser start, and no way of updating.

Interestingly the Yuzu blocker seems to have very low priority. When loading a large pages on my ancient S4 Mini, the check whether a request should be blocked can take up to half a second.
But when there is no other work to do (just checking a list of requests), it takes about 1 ms per request.
Comparing with my implementation in this PR: just checking a list of requests also takes about 1 ms per request, and when loading a page this can go up to 50 ms.
Not sure if this is relevant though... I guess if the CPU is fully loaded, it doesn't matter so much which part gets slowed down as long as it isn't the UI.

I'll update the PR a bit later, together with a bunch of comments and questions...

Helium314 commented:
I'm still looking through the code, trying to decide which parts of the Yuzu-blocker are needed, which can be removed and what needs to be replaced.
What is still needed depends a bit on how the blocker is used:

The filters are very flexible and could be used for user-defined blocking, for example:
* block all scripts
* block all scripts on google.com
* block all 3rd-party scripts
* block all 3rd-party images, but not on google.com
* block all scripts on amazon.com, but not from amazon-adsystem.com

So the adblocker could be used as a more generic content-blocker not limited to ads. But the more general use clashes with the way ad-blocking is implemented in Fulguris/Lightning: if the ad-blocker is switched off, all blocking is off.
Of course the user-defined blocking could be implemented in some different place, basically duplicating adblocker functionality with different blocklists.
But since I don't really like this kind of duplication, I am thinking of complely disabling/removing the current BloomFilterAdBlocker. The Yuzu blocker can also act as a simple hosts-blocker.
Would this be ok @Slion, or do you want to keep the current adblocker as an alternative?

Slion commented:
That's a tough question. While there is no need to keep redundant systems, during development and transition phase it can be convenient to be able to switch between the two systems. However if the new system performs just as efficiently as the old system while performing the same tasks then it should indeed replace the old system in production release.

Helium314 commented:
Allright, so for now I'll keep using the existing AdBlocker interface and add some functions. When using the BloomFilterAdBlocker they'll just do nothing.

Helium314 commented:
Finally some semi-usable version...
Blocking works, element hiding maybe (did not really test, maybe necessary js is injected too late).
Blocklists still can't be changed, but easylist and easyprivacy are loaded and updated from website
The original adblocker is still in the code, but currently not accessible.


What is left to do before it can be merged (no particular order):
* UI for adding/removing/enabling blocklists
* load lists in background. currently it's delaying browser start
* test whether element hiding works (if it does, add setting in UI)
* allow user defined filters (including whitelisting pages from adblocking) -> add UI and persist somewhere
* remove unused leftover parts from yuzu

And maybe:
* add malware and mining protection lists (and setting in UI)
* somehow use bloom filter: could accelerate the tag matching, not sure whether it's worth the work
* allow/understand hosts lists
* add some proper auto-update (currently checked only on browser start)
* remove easylist from assets, can be downloaded anyway
* add some check for duplicate filters. currently, having duplicate blocklists means blocking at half speed
* change blocklist (`AbpEntity`) storage. In yuzu it used some _Room_ DB which wasn't able to get working, so I switched to shared preferences. Since they're few and only loaded at app start, this somewhat inefficient implementation should not matter.

Helium314 commented:
**Crude summary of how the blocker is working (also for myself):**

`AbpFilterDecoder` generates filters from lines of the blocklist and sorts them into categories:
black: blocklist
white: blocklist exclusions
elementDisableFilter: element filter exclusions
elementFilter: elements to hide based on page content

These lists are lists of `UnifiedFilter`s (abstract class for different types of filters)
From these lists (4 lists for each blocklist), `FilterContainer`s are created, one for each of _black_, _white_, _elementDisableFilter_ (_elementFilter_ is handled separately, because it's not based on the web request)
This is done via something like `FilterContainer().also{ filterList.forEach(it::plusAssign) }` (I guess here would be the best point if we want to remove duplicate/unnecessary filters that may occur when using multiple blocklists)

A `FilterContainer` is basically a `HashMap<String, ContentFilter>`, where `ContentFilter` can actually contain several filters (need to have another look).
`plusAssign` creates a _best tag_ from the filter to be added, which is basically the longest string of all substrings of the filter consisting of numbers, letters and % (and at least 3 characters), and used as the key of the `FilterContainer`
E.g. _google.com/manyads_ would yield the tag manyads
If the filter is a regex or no proper tag can be created, the tag will be an empty string.

Filter matching is simply done by `FilterContainer[ContentRequest]`, which returns the matching filter or null.
Internally it works by creating _tag_s for the requested url to be checked (tags like above, but not just the "best" tag)
E.g. google.com/manyads would yield: google, com, manyads
The request is then checked against all filters that have any of those tags as _best tag_ (and against filters with empty tag)

Slion commented:
@Helium314 I really appreciate you working on this feature. I like how you document your code and the abundance of comments here. I still have not taken the time to do a proper code review. However I'm learning from your work as you go.

It seems to me that you have decided to port the Yuzu implementation then.
Remind me again, is that using a reckonized technology such as ublock or easylist?

Helium314 commented:
> It seems to me that you have decided to port the Yuzu implementation then.

Yes, I found that it is somewhat faster than my partial implementation, but already understands the full set of rules.
Unfortunately I found it rather hard to understand what is happening, which is due to the sometimes convoluted use of the many files/classes, and mostly due to complete lack of comments.

>Remind me again, is that using a reconized technology such as ublock or easylist?

Yes, it's using the [AdblockPlus ruleset](https://adblockplus.org/en/filter-cheatsheet), which I think is the most common format besides simple hosts lists.
uBlockOrigin uses/understands the same format (and has some additional filter rules for its "internal" list)

Helium314 commented:
It took a while to add some UI for the lists, and I don't really like it, but a least it sort of works...

In ad block settings, it is now possible to check available lists, modify the title and url, and enable/disable each list.
New lists can also be added.
![Screenshot_20210616-101706_Fulguris](https://user-images.githubusercontent.com/43007630/122186750-a3941180-ce8e-11eb-9e73-8f23a94d12a4.png)

Some problems/work necessary here:
Changes to the lists (and newly added lists) are not shown immediatly, it is necessary to re-enter the ad block settings
I wasn't able to add the _enabled_-switch next to the list name (where I think it should be), so it's in the edit-menu
_add list_ button should probably look a bit different than a normal list (no idea how to add a '+' icon)
there is no _delete_ option
there is no _update/refresh list_ option (update-check is currently done on browser start)
last updated time is always null, need to have a look at how this is handled in yuzu
urls are not checked for validity
not possible to use a file instead of a url
the internal list url can be changed (probably shouldn't even be displayed)

I have no experience working with UI stuff, and even the current state took a lot longer than I think it should.
So any help or hints are highly appreciated.

Helium314 commented:
Yes, definitely looks better now.
UI is mostly done and working, only problem is that it's still not possible to use a local file.

Other than that I think this PR is most ready for review

What is not working properly:
* use local file as blocklist source: needs to be implemented
* user defined filters: not yet, but not necessary to have the blocker working. Although at least a simple user-defined whitelist should be added for now
* list loading: currently still blocks the UI (delays start for 0.5-1 s on my old phone). Loading in background is much slower, so ad blocking is only ready after a few seconds (after the first page has already loaded). Not sure what's best here
* old adblocker: still present (also in UI), but not working. Probably should at least be removed from UI
* element hiding: definitely does not work, and I am not really sure whether it actually works in Yuzu browser. Maybe I'll try to implement it later
* auto-update: only on browser start. So if the browser is left open for several days, the blocklists might get outdated. Currently I feel like improvements are not worth the work
* malware/mining protection: should work, but currently there are no lists provided. Can be added later

jamal2362 commented:
@Helium314 

Thank you for the work that has been done so far.

As you said, the local files and a whitelist/blacklist are still missing. Everything else looks promising.

With the auto-update, I noticed one thing. Only the first list is updated and the others are not. You have to press the button to update all of them.

We are getting closer to the goal. There is not much more. 

Thanks again for rewriting the adblocker!

Slion commented:
> Other than that I think this PR is mostly ready for review

I went quick over it on GitHub. I like the fact that you took a lot of the files and classes directly from Yuzu.
It seems like the changes to Fulguris code are minimal though I still need to do a thorough review.

> * use local file as blocklist source: needs to be implemented
I guess we could do without it to start with and add that later.

> * user defined filters: not yet, but not necessary to have the blocker working. Although at least a simple user-defined whitelist should be added for now
I guess we could add that later too.

> * list loading: currently still blocks the UI (delays start for 0.5-1 s on my old phone). Loading in background is much slower, so ad blocking is only ready after a few seconds (after the first page has already loaded). Not sure what's best here
>   _Update_: on my somewhat faster other phone (quad instead of dual core, same per-core speed) loading in background is as fast as when blocking the UI. Assuming most phones have more than 2 cores, loading in background is probably they best way.
Background loading is certainly the way to go but we should delay loading web pages until the blocker is fully functional.

> * old adblocker: still present (also in UI), but not working. Probably should at least be removed from UI
Well I guess if you see no advantage in keeping it we should just discard it? 
Or does it make sense to be able to switch between the two systems?

> * element hiding: definitely does not work, and I am not really sure whether it actually works in Yuzu browser. Maybe I'll try to implement it later
Not needed for first release.

> * auto-update: only on browser start. So if the browser is left open for several days, the blocklists might get outdated. Currently I feel like improvements are not worth the work
Yes that should do for now. Apps on Android are restarted often enough. Though I guess we should have an option to disable auto-update.

> * malware/mining protection: should work, but currently there are no lists provided. Can be added later
Fine by me.

jamal2362 commented:
@Helium314 

Gladly, am always ready for testing :)
That makes sense, thanks for the explanation!

Helium314 commented:
Per-page whitelist now has a UI (simple checkbox in popup menu) and works.
(and I got reminded of why I really wanted to avoid working with databases on Android...)

So next things:
* remove old adblocker from UI
* test how fast user blocklists and DB interactions are with a high but realistic number of entries, like 1000 (optimize if necessary)
* clean up a bit, could be combined with review

jamal2362 commented:
I think the old adblocker should be completely removed from the code, because the new adblocker is far superior to the old adblocker. The hosts can still be so good, but adguard will always be better. Everything the normal adblocker can do, the new one can do too and much more.

But this is just my opinion.

Slion commented:
> actually think some way of disabling blocking for specific pages/domains is really useful to have

Definitly but if not having it helps us bring the core of the changes to production quality faster then we could do without it at first. It will also help to have smaller more focused changes to review.

Helium314 commented:
> test how fast user blocklists and DB interactions are with a high but realistic number of entries, like 1000 (optimize if necessary)

On my old phone loading a 2400 entry list takes about 1.7s when done at the same time as loading ad lists, half time when doing after loading is finished. Definitely should be optimized, but considering that most people have a faster phone and much(!) shorter list, I think it's ok.

> Definitly but if not having it helps us bring the core of the changes to production quality faster then we could do without it at first. It will also help to have smaller more focused changes to review.

well, it's already implemented anyway :smile: 

I will mark the PR as ready for review now.
There are still some things that should be done/decided, but this is mentioned in comments.
Then there are smaller things like consistent and clear naming of variables/classes, making some functions/variables private, removing more unused code, adjustment of used strings. These should probably be part of the review, often I am not sure what is best.

Slion commented:
@Helium314 I just pushed significant changes to the main branch so you may want to do a rebase, let me know if you need help with that.

I also added you as a collaborator to that repository so you should be able to push those changes into a development branch. Maybe name it dev-content-blocker.

I created issue #213 to document and track progress. I'm hoping that being a collaborator gives you the right to edit my description of that issue. If not then just do a post of your own.

Again if I were you I would focus on minimal UI and functionality set for an easier to test and more focus initial contribution and then build on that once released to production.

Next step as I see it:
* Rebase and push to dev-content-blocker,
* Test and Code review and repeat until satisfaction,
* Push to main.
* Internal release on Google Play, I'll need to add you to internal testers, need your e-mail address.
* More testing.
* Public release once happy.

Planning for that feature to be released next certainly as part of 1.7.x

Helium314 commented:
I'll do/try that a bit later... don't really have time to work on this until at least tomorrow evening (and just tried to have it ready for review by today noon, i.e. now)

One potential problem that just came to my mind is updating from an old version. Never tried that, and not 100% sure it will work.

jamal2362 commented:
We should also not forget the option with the local file.
This is also still open. 😀

Slion commented:
> I'll do/try that a bit later... don't really have time to work on this until at least tomorrow evening (and just tried to have it ready for review by today noon, i.e. now)

Of course no pressure take all the time you need.

> One potential problem that just came to my mind is updating from an old version. Never tried that, and not 100% sure it will work.

What do you mean? Rebase and conflict resolution? I could do that for you if you want.

Slion commented:
@Helium314 I've rebased and pushed those changes there https://github.com/Slion/Fulguris/pull/214

Still have to do a formal code review though.
As far as I could tell it still works fine, hopefully I have not broken anything in the process.
If you are ok with that you should work on that new branch from now on.
Closing this pull request as it is superseded by the above.

Helium314 commented:
Great, thank you!
Only now back at my PC again, will update local code and do a quick check.


> > One potential problem that just came to my mind is updating from an old version. Never tried that, and not 100% sure it will work.
> 
> What do you mean? Rebase and conflict resolution? I could do that for you if you want.

I mean updating from an old (or current) version of Fulguris to one with the new blocker. Maybe some specific combination of user settings could cause trouble? But looking through the the code, no old adblock settings except on/off are used in the new blocker.

Helium314 commented:
> Internal release on Google Play, I'll need to add you to internal testers, need your e-mail address.

I don't use play services on my phone, and have actually not used the Play store even once.
I would prefer to keep it that way, but if you like I can set up my old test phone with google services / get some account there.

Slion reviewed (commented):
> > Internal release on Google Play, I'll need to add you to internal testers, need your e-mail address.
> 
> I don't use play services on my phone, and have actually not used the Play store even once.
> I would prefer to keep it that way, but if you like I can set up my old test phone with google services / get some account there.

That's ok we can do without Google play.

Helium314 commented on app/src/main/java/acr/browser/lightning/preference/UserPreferences.kt:
```diff
@@ -406,12 +407,12 @@ class UserPreferences @Inject constructor(
      * User can disable Firebase Google Analytics.
      */
-    var analytics by preferences.booleanPreference(R.string.pref_key_analytics, R.bool.pref_default_analytics)
+    var analytics by preferences.booleanPreference(R.string.pref_key_analytics, true)
```
no, definitely not relevant for the blocker

Helium314 commented on app/src/main/java/acr/browser/lightning/preference/UserPreferences.kt:
```diff
@@ -406,12 +407,12 @@ class UserPreferences @Inject constructor(
      * User can disable Firebase Crash Report AKA Crashlytics.
      */
-    var crashReport by preferences.booleanPreference(R.string.pref_key_crash_report, R.bool.pref_default_crash_report)
+    var crashReport by preferences.booleanPreference(R.string.pref_key_crash_report, true)
```
also definitely not relevant for the blocker
